###########################################################################################
# Script for the DESEq2 analysis after featureCounts & annotations
#
# Generated by: Dilek Cansu GURER, Dec 2021
# Usage: Rscript deseq2_script.R
###########################################################################################

# Import required libraries
library(DESeq2)
library(dplyr)
library(ggplot2)
library(clusterProfiler)
library(biomaRt)
library(ReactomePA)
library(DOSE)
library(KEGG.db)
library(org.Hs.eg.db)
library(pheatmap)
library(genefilter)
library(RColorBrewer)
library(GO.db)
library(topGO)
library(dplyr)
library(gage)
library(ggsci)
library(tidyverse)
library(rtracklayer)

# Change working dir to where you store your count matrix file 
setwd("C:/Users/dcans/Desktop/r_calisma/cp_mrnaseq/")

# Import gene counts table
# Skip first row (skip=1)
# Make row names gene identifiers
countdata <- read.table("final_counts.txt", header = TRUE, skip = 1, row.names = 1)

# Remove .bam + '..' from column identifiers
colnames(countdata) <- gsub(".bam", "", colnames(countdata), fixed = T)
colnames(countdata) <- gsub("..", "", colnames(countdata), fixed = T)

# Remove length/char columns
countdata <- countdata[ ,c(-1:-5)]

# Import metadata file
# Make row names the matching sampleID's from the countdata
metadata <- read.delim("metadata.txt", row.names = 1)

# Add sampleID's to the mapping file
metadata$sampleid <- row.names(metadata)

# Reorder sampleID's to match featureCounts column order. 
metadata <- metadata[match(colnames(countdata), metadata$sampleid), ]

# Make sure ID's are correct
head(metadata)

# countData : count dataframe
# colData : sample metadata in the dataframe with row names as sampleID's
# design : The design of the comparisons to use. 
#            Use (~) before the name of the column variable to compare
ddsMat <- DESeqDataSetFromMatrix(countData = countdata,
                                 colData = metadata,
                                 design = ~condition)

# Find differential expressed genes
ddsMat <- DESeq(ddsMat)

# Get results from testing with FDR adjust pvalues
results <- results(ddsMat, pAdjustMethod = "fdr", alpha = 0.05)

# Generate summary of testing. 
summary(results)

# Check directionality of the log2 fold changes
## Log2 fold change is set as (CP / DMSO)
## Postive fold changes = Increased in CP
## Negative fold changes = Decreased in CP
mcols(results, use.names = T)

# Add gene full name
# First generate GTF table with ENSEMBL gene IDs, biotypes and gene names
gtf <- import("gencode.v38.primary_assembly.annotation.gtf")%>%
  as_tibble %>%
  distinct(gene_id, gene_name, gene_type)

# Check if we can correctly generate GTF table
head(gtf)

# Generate another column with ENSEMBL gene IDs in results matrix to be able to use filter function with ids vector
# Check if we can add IDs column correctly
results$ids <- row.names(results)
head(results)

# Filter imported GTF data
# Add biotypes from this filter to results
# Agg gene names from this filter to results
res_biotype <- dplyr::filter(gtf, gene_id %in% results$ids)
head(res_biotype)
results$biotype <- res_biotype$gene_type
head(results)
results$gene_name <- res_biotype$gene_name


# Remove strings after "." in ENSENBL gene IDs
results$ids <- gsub("\\..*","",results$ids)

# Add gene full name
results$description <- mapIds(x = org.Hs.eg.db,
                              keys = results$ids,
                              column = "GENENAME",
                              keytype = "ENSEMBL",
                              multiVals = "first")

# Add ENTREZ ID
results$entrez <- mapIds(x = org.Hs.eg.db,
                                  keys = results$ids,
                                  column = "ENTREZID",
                                  keytype = "ENSEMBL",
                                  multiVals = "first")

# Select only significant genes
results_sig <- subset(results, padj < 0.05)
head(results_sig)

# Write significant results table to .txt file
write.table(x = as.data.frame(results_sig), 
            file = "results_gene_annotated_significant_biotypes.txt", 
            sep = '\t', 
            quote = F,
            col.names = NA)

# Write normalized gene counts to a .txt file
write.table(x = as.data.frame(counts(ddsMat), normalized = T), 
            file = 'normalized_counts_all.txt', 
            sep = '\t', 
            quote = F,
            col.names = NA)

# Write significant normalized gene counts to a .txt file
write.table(x = counts(ddsMat[row.names(results_sig)], normalized = T), 
            file = 'normalized_counts_significant.txt', 
            sep = '\t', 
            quote = F, 
            col.names = NA)

# Write annotated results table to a .txt file (including non-significants)
write.table(x = as.data.frame(results), 
            file = "results_including_non-significants.txt", 
            sep = '\t', 
            quote = F,
            col.names = NA)
